### DISCUSSION TITLE
Build caching not working

### DISCUSSION THREAD

**Discussion by scott-coates** [vercel/community#222] [Help] [answered]
I am developing a monorepo app in next.js. My builds routinely take 3-4 minutes. Most of this time is wasted by re-downloading npm packages.

![image](https://user-images.githubusercontent.com/1829591/150692855-18c28686-6e23-4289-ac7a-bd3054b4b7f0.png)

According to vercel docs:  https://vercel.com/docs/concepts/deployments/build-step#caching
> Vercel intelligently caches files based on the Framework Preset selected in your Project Settings. The following files are cached in addition to node_modules/**, yarn.lock, and package-lock.json:

Why aren't my dependencies within `node_modules` being cached during a vercel deploy?

**Comment by jaredpalmer** [comment, marked as answer]
You can set `ENABLE_ROOT_PATH_BUILD_CACHE` as `1` an environment variable in the dashboard or in `vercel.json` and this will be solved. We are working on automatically enabling this feature when workspaces are detected

**Reply by klapec** [reply]
Has anything changed in this matter? I'm seeing lots of time burned on `yarn install`-s for my monorepo apps even though I have the `ENABLE_ROOT_PATH_BUILD_CACHE` env set up.

**Reply by valeriangalliat** [reply]
Same here, despite setting `ENABLE_ROOT_PATH_BUILD_CACHE=1`, `yarn install` has to re-fetch everything on every build in my monorepo

Edit: looks like Yarn3 re-fetches everything despite an up-to-date `node_modules` being there if it doesn't find its `.yarn/cache` folder, which is not cached by Vercel by default

By setting my install command to `YARN_CACHE_FOLDER=../../node_modules/.yarn-cache yarn workspaces focus` I got full cached installed to work!

**Reply by JClackett** [reply]
ENABLE_ROOT_PATH_BUILD_CACHE doesn't seem to do anything, using yarn workspaces + nextjs

**Reply by pepijn-vanvlaanderen** [reply]
@JClackett `ENABLE_ROOT_PATH_BUILD_CACHE` is enabled by default when Vercel [detects workspaces](https://vercel.com/changelog/faster-build-times-for-monorepos).

@valeriangalliat did you have any specific `.yarnrc.yml` options set as well? Because it only seems to work with existing cache. Whenever a clean build is triggered, an error is thrown by Yarn, i.e.:
```
 YN0001: │ Error: While persisting /vercel/path0/node_modules/.yarn-cache/@babel-core-npm-7.21.0-fb3817b0e5-357f4dd363.zip/node_modules/@babel/core/ -> /vercel/path0/node_modules/@babel/core ENOENT: no such file or directory, scandir '/vercel/path0/node_modules/.yarn-cache/@babel-core-npm-7.21.0-fb3817b0e5-357f4dd363.zip/node_modules/@babel/core'
```

**Reply by JClackett** [reply]
> @JClackett `ENABLE_ROOT_PATH_BUILD_CACHE` is enabled by default when Vercel [detects workspaces](https://vercel.com/changelog/faster-build-times-for-monorepos).
> 
> 
> 
> @valeriangalliat did you have any specific `.yarnrc.yml` options set as well? Because it only seems to work with existing cache. Whenever a clean build is triggered, an error is thrown by Yarn, i.e.:
> 

Hmm, definitely not caching it though, using yarn and workspaces

**Reply by valeriangalliat** [reply]
I've had some issues with setting the yarn cache folder in node_modules as well, but more recently I found https://github.com/vercel/turbo/issues/785#issuecomment-1060054306 and switched to `YARN_CACHE_FOLDER=.next/cache/yarn-cache` and I've had better results with this so far!

**Reply by ghost** [reply]
@valeriangalliat Do you use a turborepo setup? If so, do you think it's turbo who caches the yarn cache for you? I tried the same thing in my turborepo setup and it doesn't seems to work. Looks like `.next/cache` is ignored in the turbo build output caching step.

**Comment by pepijn-vanvlaanderen** [comment]
@jaredpalmer it seems (at least with the latest yarn v3) to not be possible to set the cache folder of yarn inside the node_modules folder and since ENABLE_ROOT_PATH_BUILD_CACHE. 

It seems to be a quick resolution if Vercel adds caching for the default yarn cache folder, which is `.yarn/cache`, is this something that can be included in the roadmap? Or are there any other folders cached which we can point the yarn cache to?
